### ISSUE TITLE
PWAs

### ISSUE THREAD

**Issue by arctic-hen7** [arctic-hen7/perseus#71]
**Is your feature request related to a problem? Please describe.**
Progressive Web Apps (PWAs) have become incredibly powerful over the last few years, and can even work as viable Android apps now (see [here](https://web.dev/using-a-pwa-in-your-android-app/)). Right now, Wasm PWAs are practically nonexistent except for through Blazor, and Perseus has a significant opportunity to work with PWAs out of the box.

**Describe the solution you'd like**
This would likely be achieved with a plugin that adds PWA functionality. It would need to add the necessary manifest details, a few static assets, and a service worker. The last part is the hardest, because Wasm and service workers don't really work together yet, so this would probably have to be in JS for now. Nonetheless, it should be possible to cache the main Wasm binary and HTML/JSON resources without any changes to the Perseus server. However, E-Tag support on the server is long overdue, and this would improve cacheability and thus performance for non-exported apps.

**Describe alternatives you've considered**
A version of Perseus for apps, but that would require a ridiculous amount of work and would inevitably end up leading to fractured codebases.

**Additional context**
This is a tracking issue for PWAs in Perseus, and this should be a place for ideas on the topic as well as development updates. As I said, nearly all this work will occur in a plugin `perseus-pwa`, but there may also be some changes needed in the core.

**Comment by afidegnum** [comment]
I have come across this, I don't know something can be useful from it. 

[This](https://woz.sh/) and this  [PWA](https://hub.packtpub.com/introducing-woz-a-progressive-webassembly-application-pwa-web-assembly-generator-written-entirely-in-rust/)

**Comment by arctic-hen7** [comment]
Hmm okay, that looks really interesting, thanks! Woz seems to be archived now, but it's certainly something to look into for ideas and possible optimizations so we don't have to reinvent the wheel.

**Comment by erlend-sh** [comment]
> The last part is the hardest, because Wasm and service workers don't really work together yet, so this would probably have to be in JS for now.

@dunnock might have input on this, having built https://github.com/dunnock/wasi-worker

**Comment by arctic-hen7** [comment]
Hmm that could be extremely useful, thanks!

**Comment by afidegnum** [comment]
But isn't perseus a kind of PWA on its own coupled with sycamore ?

**Comment by arctic-hen7** [comment]
No, a PWA is installable; the name is a bit misleading (PWAs are actually a specific type of web app with certain extra metadata). This would allow Perseus to be used for the development of what are essentially mobile apps, just running in a WebView or similar. For example, PWAs can use APIs for music player controls on your phone's lockscreen, which would usually be gated behind the browser.

**Comment by afidegnum** [comment]
ok, still exploring possibilities.

**Comment by arctic-hen7** [comment]
Just a brief update on this front: once #173 is done, Perseus will be able to make pages interactive from a single document request, along with the Wasm and JS bundles, which should make future PWA support snappier. In terms of implementation plans, I'll aim to start exploring how exactly I want this to work (and how this would affect the core codebase) over the weekend.
